File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/media/files/properties/playstats/UpdatePlayStatsOnCompleteRegistration.java
package com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.playstats;

import android.content.BroadcastReceiver;
import android.content.IntentFilter;

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.FilePropertiesStorage;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.SessionFilePropertiesProvider;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.playstats.factory.PlaystatsUpdateSelector;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.repository.FilePropertyCache;
import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.client.connection.receivers.IConnectionDependentReceiverRegistration;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents;
import com.lasthopesoftware.bluewater.client.servers.version.ProgramVersionProvider;
import com.lasthopesoftware.resources.scheduling.ParsingScheduler;

import java.util.Collection;
import java.util.Collections;

public class UpdatePlayStatsOnCompleteRegistration implements IConnectionDependentReceiverRegistration {

	private static final Collection<IntentFilter> intents = Collections.singleton(new IntentFilter(PlaylistEvents.onFileComplete));

	@Override
	public BroadcastReceiver registerWithConnectionProvider(IConnectionProvider connectionProvider) {
		final FilePropertyCache cache = FilePropertyCache.getInstance();

		return new UpdatePlayStatsOnPlaybackCompleteReceiver(
			new PlaystatsUpdateSelector(
				connectionProvider,
				new SessionFilePropertiesProvider(connectionProvider, cache, ParsingScheduler.instance()),
				new FilePropertiesStorage(connectionProvider, cache),
				new ProgramVersionProvider(connectionProvider)));
	}

	@Override
	public Collection<IntentFilter> forIntents() {
		return intents;
	}


}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/media/files/properties/playstats/UpdatePlayStatsOnCompleteRegistration.kt
package com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.playstats

import android.content.IntentFilter
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.ScopedFilePropertiesProvider
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.playstats.factory.PlaystatsUpdateSelector
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.repository.FilePropertyCache
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.properties.storage.ScopedFilePropertiesStorage
import com.lasthopesoftware.bluewater.client.browsing.library.revisions.ScopedRevisionProvider
import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider
import com.lasthopesoftware.bluewater.client.connection.authentication.ScopedConnectionAuthenticationChecker
import com.lasthopesoftware.bluewater.client.connection.receivers.RegisterReceiverForEvents
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.messages.PlaylistMessages.TrackCompleted
import com.lasthopesoftware.bluewater.client.servers.version.ProgramVersionProvider
import com.lasthopesoftware.bluewater.shared.android.messages.ReceiveBroadcastEvents
import com.lasthopesoftware.bluewater.shared.cls
import com.lasthopesoftware.bluewater.shared.messages.application.ApplicationMessage

class UpdatePlayStatsOnCompleteRegistration : RegisterReceiverForEvents {
	override fun registerBroadcastEventsWithConnectionProvider(connectionProvider: IConnectionProvider): ReceiveBroadcastEvents {
		return ReceiveBroadcastEvents {  }
	}

	override fun registerWithConnectionProvider(connectionProvider: IConnectionProvider): (ApplicationMessage) -> Unit {
		val cache = FilePropertyCache.getInstance()
		val scopedRevisionProvider = ScopedRevisionProvider(connectionProvider)
		return UpdatePlayStatsOnPlaybackCompleteReceiver(
			PlaystatsUpdateSelector(
				connectionProvider,
				ScopedFilePropertiesProvider(connectionProvider, scopedRevisionProvider, cache),
				ScopedFilePropertiesStorage(
					connectionProvider,
					ScopedConnectionAuthenticationChecker(connectionProvider),
					scopedRevisionProvider,
					cache
				),
				ProgramVersionProvider(connectionProvider)
			)
		)
	}

	override fun forIntents(): Collection<IntentFilter> = emptySet()

	@Suppress("UNCHECKED_CAST")
	override fun forClasses(): Collection<Class<ApplicationMessage>> = classes as Collection<Class<ApplicationMessage>>

	companion object {
		private val classes = setOf<Class<*>>(cls<TrackCompleted>())
	}
}
